### PR TITLE
Auth API routes

### LINKED ISSUE #6: API Authorisation & Authentication
As a frontend engineer, I would like to be able to authenticate with the API.

- [x] Username/Email & password Authentication Route(s) with Bcrypt that provides a JWT
- [x] Validation for authentication routes
- [x] JWT middleware to maintain session
- [x] A middleware to check if the request is authorized
- [x] Test the endpoints

### PR DESCRIPTION
**Pull request by mosesokemwa**
The auth route supports registration and user login.
User login return a JWT token
Middleware to check if a request is authorised

closes #6

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
 .DS_Store
 node_modules
+db.js
```
Comment (Jakeii) [inline .gitignore]: this is already in server/.gitignore

File: server/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+    "env": {
+        "browser": true,
```
Comment (Jakeii) [inline server/.eslintrc.js]: This should be `node: true` to validate node.js rather than `browser: true`

File: server/.gitignore [hunk 3]
```diff
@@ -1,3 +1,107 @@
-node_modules
+# Logs
```
Comment (Jakeii) [inline server/.gitignore]: where did all this come from?

File: server/routes/auth.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    prefix: '/auth'
+});
+
+
+/**
+ * You can register with:
+ * curl -X POST -d 'user[email]=kenya@tunapanda.org&user[username]=mountkenya&user[hash]=mountkenya' http://localhost:3000/api/v1/auth/register
+ * curl -X POST -d '{ "user": { "username": "okemwa", "hash": "okemwaokewamaokemwa", "email":"moses@okemwa.org" } }' http://localhost:3000/api/v1/auth/register
+ */
+router.post('/register', validateAuthRoutes.validateNewUser, getUserByUsername, async ctx => {
```
Comment (Jakeii) [inline server/routes/auth.js]: Although it is authentication related, in order to maintain the CRUD convention, user creations should be via POST to `/users`

File: server/db/queries.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+const knex = require('./db');
```
Comment (Jakeii) [inline server/db/queries.js]: As we're using an ORM I think having a separate queries file is a little redundant?

**Review by Jakeii — CHANGES_REQUESTED** [review]
Great, just a few changes needed.

File: server/.gitignore (hunk 3 above)
Comment (mosesokemwa) [reply server/.gitignore]: I used [gitignore.io](gitignore.io) which generates it from open-source templates

File: server/db/queries.js (hunk 5 above)
Comment (mosesokemwa) [reply server/db/queries.js]: I see your point, also the queries in this file are not used anywhere. Let me remove it to clean up the files.

File: server/routes/auth.js (hunk 4 above)
Comment (mosesokemwa) [reply server/routes/auth.js]: Mick made these changes. They will be included in the next push to this branch

File: .gitignore (hunk 1 above)
Comment (mosesokemwa) [reply .gitignore]: Fixed

File: server/.gitignore (hunk 3 above)
Comment (mosesokemwa) [reply server/.gitignore]: I used [gitignore.io](gitignore.io) which generates it from open-source templates

**Pull request closed without merging**